File: backend/app/models/quotation.py
from datetime import date, datetime

from sqlalchemy import DATE, DECIMAL, Boolean, DateTime, Enum, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.customer import Customer
from app.models.deal import Deal
from app.models.mixins import SoftDeleteMixin, TimestampMixin
from app.models.product import Product
from app.models.user import BigPK

QUOTATION_STATUSES = ("draft", "sent", "accepted", "rejected", "expired", "converted")

# Which of the admin's two (see doc_template_service.LANGUAGES) uploaded
# quotation templates this quotation was raised in -- set at creation
# (Sales' choice on the "New quotation" form) and used as the default for
# Print/Email; a caller can still request the other language's PDF via
# the `language` query param without changing this.
QUOTATION_LANGUAGES = ("en", "ar")

# Status transitions allowed from each current status. Used by the service
# layer to reject invalid jumps (e.g. draft -> converted directly).
ALLOWED_TRANSITIONS = {
    "draft": {"sent", "rejected"},
    "sent": {"accepted", "rejected", "expired"},
    "accepted": {"converted"},
    "rejected": set(),
    "expired": set(),
    "converted": set(),
}

# Sales closing a quotation without generating an order must supply a
# reason (see quotation_service.change_status); 'rejected' is the only
# manually-driven terminal-without-order status -- 'expired' is a
# calendar-driven state, not a deliberate close.
STATUSES_REQUIRING_CLOSE_REASON = {"rejected"}


class Quotation(Base, TimestampMixin, SoftDeleteMixin):
    __tablename__ = "quotations"

    id: Mapped[int] = mapped_column(BigPK, primary_key=True)
    quotation_number: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    customer_id: Mapped[int] = mapped_column(BigPK, ForeignKey("customers.id"), nullable=False)
    # The deal this quotation belongs to (see models/deal.py) -- inherited
    # from its feasibility check if it has one, otherwise newly minted.
    deal_id: Mapped[int | None] = mapped_column(BigPK, ForeignKey("deals.id"), nullable=True)
    quotation_date: Mapped[date] = mapped_column(DATE, nullable=False)
    valid_until: Mapped[date | None] = mapped_column(DATE, nullable=True)
    status: Mapped[str] = mapped_column(
        Enum(*QUOTATION_STATUSES, name="quotation_status"), nullable=False, default="draft"
    )
    language: Mapped[str] = mapped_column(
        Enum(*QUOTATION_LANGUAGES, name="quotation_language"), nullable=False, default="en"
    )
    subtotal_amount: Mapped[float] = mapped_column(DECIMAL(14, 2), nullable=False, default=0)
    # Percentage, e.g. 0 or 10 -- a whole-document discount applied on
    # top of the already line-discounted subtotal.
    discount_percent: Mapped[float] = mapped_column(DECIMAL(5, 2), nullable=False, default=0)
    discount_amount: Mapped[float] = mapped_column(DECIMAL(14, 2), nullable=False, default=0)
    total_amount: Mapped[float] = mapped_column(DECIMAL(14, 2), nullable=False, default=0)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    converted_order_id: Mapped[int | None] = mapped_column(BigPK, nullable=True)
    # Every quotation must originate from a passed/exception-approved
    # feasibility check (enforced in quotation_service.create_quotation).
    feasibility_id: Mapped[int | None] = mapped_column(
        BigPK, ForeignKey("feasibility_checks.id"), nullable=True
    )
    # True when the system drafted this automatically because a
    # feasibility check just passed (see feasibility_service.py's
    # auto-creation hook), false for a person-created quotation. Purely
    # informational -- an auto-created quotation is a completely normal,
    # editable/deletable draft otherwise.
    auto_created: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    # Set when Sales closes this quotation without converting it to an order.
    close_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    # A quotation whose discount (document-level or any single line's) is
    # at/above Settings -> large_discount_approval_threshold can't leave
    # 'draft' until an admin approves it.
    approved_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    approved_by: Mapped[int | None] = mapped_column(BigPK, ForeignKey("users.id"), nullable=True)
    # Set when this quotation was created (or last edited) despite a raw
    # material it needs also being needed by another still-open
    # quotation/order, and Sales explicitly acknowledged that overlap --
    # see quotation_service.check_material_conflicts. A snapshot of what
    # was flagged at that moment, not a live check (availability moves on).
    material_conflict_acknowledged: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    material_conflict_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    customer: Mapped[Customer] = relationship(lazy="joined")
    deal: Mapped[Deal | None] = relationship(lazy="joined")
    lines: Mapped[list["QuotationDetail"]] = relationship(
        back_populates="quotation",
        cascade="all, delete-orphan",
        order_by="QuotationDetail.id",
    )


class QuotationDetail(Base):
    __tablename__ = "quotation_details"

    id: Mapped[int] = mapped_column(BigPK, primary_key=True)
    quotation_id: Mapped[int] = mapped_column(
        BigPK, ForeignKey("quotations.id", ondelete="CASCADE"), nullable=False
    )
    product_id: Mapped[int] = mapped_column(BigPK, ForeignKey("products.id"), nullable=False)
    quantity: Mapped[float] = mapped_column(DECIMAL(14, 4), nullable=False)
    unit_price: Mapped[float] = mapped_column(DECIMAL(14, 2), nullable=False)
    # Percentage, e.g. 0 or 10 -- this line's own discount, applied
    # before the document-level discount_percent.
    discount_percent: Mapped[float] = mapped_column(DECIMAL(5, 2), nullable=False, default=0)
    line_total: Mapped[float] = mapped_column(DECIMAL(14, 2), nullable=False)

    quotation: Mapped[Quotation] = relationship(back_populates="lines")
    product: Mapped[Product] = relationship(lazy="joined")

File: frontend/src/types/quotation.ts
/** Mirrors backend/app/schemas/quotation.py. */

export type QuotationStatus = 'draft' | 'accepted' | 'rejected' | 'expired' | 'converted'
export type QuotationLanguage = 'en' | 'ar'

export interface QuotationLineInput {
  product_id: number
  quantity: number
  unit_price: number
  discount_percent?: number
}

export interface QuotationLine extends QuotationLineInput {
  id: number
  product_code: string | null
  product_name: string | null
  unit: string | null
  discount_percent: number
  line_total: number
}

export interface Quotation {
  id: number
  quotation_number: string
  customer_id: number
  customer_name: string | null
  customer_email: string | null
  deal_id: number | null
  deal_number: string | null
  feasibility_id: number | null
  /** True when the system drafted this automatically because a
   * feasibility check just passed, rather than a person creating it. */
  auto_created: boolean
  quotation_date: string
  valid_until: string | null
  status: QuotationStatus
  /** Which admin-uploaded template (Admin -> Documents) this was raised
   * in -- what Print/Email default to rendering. */
  language: QuotationLanguage
  subtotal_amount: number
  discount_percent: number
  discount_amount: number
  total_amount: number
  notes: string | null
  converted_order_id: number | null
  /** A still-relevant concern from the feasibility check this was raised
   * from (approved despite a shortfall, or the check has since been
   * revived and no longer reflects the approval this relied on). Null
   * for a standalone quotation or one whose check is a clean pass. */
  feasibility_blocker: string | null
  /** Set once an admin has approved a large discount on this quotation
   * (Settings -> large_discount_approval_threshold). Null if never
   * required or not yet approved. */
  approved_at: string | null
  /** True when this quotation's material needs overlapped another
   * still-open quotation/order and Sales explicitly proceeded anyway --
   * see check_material_conflicts. */
  material_conflict_acknowledged: boolean
  material_conflict_details: MaterialConflict[] | null
  lines: QuotationLine[]
  /** Stamped every time Sales logs a customer follow-up (a call, an
   * email outside this app) -- distinct from being emailed through
   * this app. See next_followup_date/followup_status. */
  last_followup_at: string | null
  next_followup_date: string | null
  followup_status: FollowupStatus
  /** Where this quotation stands re: becoming an order, and (for
   * 'blocked') exactly why -- see conversion_block_reasons. */
  conversion_status: ConversionStatus
  conversion_block_reasons: string[]
  created_at: string
  updated_at: string
}

export type FollowupStatus = 'not_due' | 'due' | 'overdue' | 'completed'
export type ConversionStatus = 'converted' | 'ready' | 'blocked'

export interface MaterialConflictCompetitor {
  quotation_id: number
  quotation_number: string
}

export interface MaterialConflict {
  raw_material_id: number
  code: string
  name: string
  unit: string
  required_by_this: number
  available: number
  shortfall: number
  competing_quotations: MaterialConflictCompetitor[]
}

export interface QuotationPayload {
  customer_id: number
  feasibility_id?: number | null
  quotation_date: string
  valid_until?: string | null
  notes?: string | null
  /** Percentage, e.g. 0 or 10 -- a whole-document discount on top of
   * any per-line discounts. */
  discount_percent?: number
  lines: QuotationLineInput[]
  language: QuotationLanguage
  material_conflict_acknowledged?: boolean
}

/** The only status changes a person can make. 'converted' is set by
 * create_order_from_quotation and 'expired' by the scheduled scan. */
export type SettableQuotationStatus = 'accepted' | 'rejected'
